cmd/kubeadm/app/images/images.go

```diff
@@ -48,8 +48,9 @@ func GetDNSImage(cfg *kubeadmapi.ClusterConfiguration) string {
 	if cfg.DNS.ImageRepository != "" {
 		dnsImageRepository = cfg.DNS.ImageRepository
 	}
-	// Handle the renaming of the official image from "registry.k8s.io/coredns" to "registry.k8s.io/coredns/coredns
-	if dnsImageRepository == kubeadmapiv1beta3.DefaultImageRepository {
+	// Handle the renaming of the official image from "registry.k8s.io/coredns" to "registry.k8s.io/coredns/coredns"
```

Comment:
this is actually not valid. the official image remains with a /coredns subpath. if thr user has a custom repo the subpath is always absent.

Reply:
I don't understand what's not valid here.

```
$ docker pull k8s.gcr.io/coredns:v1.8.6
Trying to pull k8s.gcr.io/coredns:v1.8.6...
Error: initializing source docker://k8s.gcr.io/coredns:v1.8.6: reading manifest v1.8.6 in k8s.gcr.io/coredns: manifest unknown: Failed to fetch "v1.8.6" from request "/v2/coredns/manifests/v1.8.6".

$ docker pull k8s.gcr.io/coredns/coredns:v1.8.6
Trying to pull k8s.gcr.io/coredns/coredns:v1.8.6...
Getting image source signatures
Copying blob sha256:88efb86cbcab356a7db0ba9ea09694d394f2e088f3e5119aa331025c1e6cb7fa
[...]

$ docker pull registry.k8s.io/coredns:v1.8.6
Trying to pull registry.k8s.io/coredns:v1.8.6...
Error: initializing source docker://registry.k8s.io/coredns:v1.8.6: reading manifest v1.8.6 in registry.k8s.io/coredns: manifest unknown: Failed to fetch "v1.8.6"

$ docker pull registry.k8s.io/coredns/coredns:v1.8.6
Trying to pull registry.k8s.io/coredns/coredns:v1.8.6...
Getting image source signatures
Copying blob sha256:88efb86cbcab356a7db0ba9ea09694d394f2e088f3e5119aa331025c1e6cb7fa
[...]
```

So `k8s.gcr.io` and `registry.k8s.io` seem to behave the same, and therefore require the same workaround. Before https://github.com/kubernetes/kubernetes/commit/50bea1dad89930ad565526910aadc314b9e9f38b, this exact code location did the rewrite for `k8s.gcr.io`. Or did I mistake something?

Reply:
this code:
https://github.com/kubernetes/kubernetes/commit/50bea1dad89930ad565526910aadc314b9e9f38b#diff-54f02c1435b3faae5f85c0507e8ccb2d3986f080287d9de55473b6587e126fa9R51
is about appending /coredns to registry.k8s.io so that the location ends up as registry.k8s.io/coredns/coredns:foo and is the valid one.
for newer versions of kubeadm we should not track the old registry k8s.gcr.io.
at that point kubeadm considers k8s.gcr.io a "custom" registry and the coredns image location is k8s.gcr.io/coredns:foo 
this is by design and was done for backwards compatibility once coredns maintainers introduced the /coredns subpath.